README.md
smstraderb
==========

# Description

This gem provides a Ruby client library for the smstrade.de HTTP API.

# Installation

    # gem install smstraderb

# Usage

    require 'smstraderb'

    key = 'you-smstrade-api-key'

    sms = SMSTradeRB.new(:route => :basic, :key => key, :debug => true)
    res = sms.send(:to => '123 456789', :message => 'hello api')

    if res.ok?
      puts "success!"
    else
      puts "error:"
      puts res.response_message
    end

# Contribute

* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but bump version
  in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.

# Copyright

Copyright (c) 2010 Bernd Ahlers. See LICENSE for details.


CHANGES.md
smstraderb - ChangeLog
======================
# 0.1.0
* Ensure escaping of the phone number and the message.
* The rack server app now provides the call params for inspection.

# 0.0.1
* Initial release.


spec/spec_helper.rb
$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
require 'rubygems'
require 'rspec'
require 'rspec/autorun'
require 'rack/test'
require 'artifice'

RSpec.configure do |config|
  include Rack::Test::Methods
end


spec/smstraderb_spec.rb
require 'spec_helper'
require 'smstraderb'
require 'smstraderb/server'

describe SMSTradeRB do
  it "sets the key if it's initialized with the :key option" do
    SMSTradeRB.new(:key => 'myapikey').key.should == 'myapikey'
  end

  it "sets the message if initialized with the :message option" do
    SMSTradeRB.new(:message => 'mymessage').message.should == 'mymessage'
  end

  describe "to option" do
    it "sets the recipient (to) if initialized with the :to option" do
      SMSTradeRB.new(:to => '123456').to.should == '123456'
    end

    it "raises an exception if the number is invalid" do
      expect{
        SMSTradeRB.new(:to => '123456abc')
      }.to raise_error(SMSTradeRB::InvalidFormat)
    end

    it "removes whitespace" do
      SMSTradeRB.new(:to => '12 34  56').to.should == '123456'
    end

    it "removes dashes (-)" do
      SMSTradeRB.new(:to => '12-34-56').to.should == '123456'
    end

    it "removes parenthesis" do
      SMSTradeRB.new(:to => '(123)456').to.should == '123456'
    end

    it "allows phone numbers with a country prefix" do
      SMSTradeRB.new(:to => '+49 12345').to.should == '+4912345'
    end
  end

  describe "from option" do
    context "route set to :gold" do
      it "allows the from option to be set" do
        SMSTradeRB.new(:route => :gold, :from => '123456').from.should == '123456'
      end
    end

    context "route set to :direct" do
      it "allows the from option to be set" do
        SMSTradeRB.new(:route => :direct, :from => '123456').from.should == '123456'
      end
    end

    it "raises an exception if route is not :gold or :direct" do
      expect {
        SMSTradeRB.new(:from => '123456').from.should == '123456'
      }.to raise_error(SMSTradeRB::InvalidOption)
    end

    it "verifies the length (max 11)" do
      expect {
        SMSTradeRB.new(:route => :gold, :from => 'ab cd sdfses')
      }.to raise_error(SMSTradeRB::InvalidFormat)
    end
  end

  describe "routes" do
    it "sets the route if initialized with the :route option" do
      SMSTradeRB.new(:route => :gold).route.should == :gold
    end

    it "sets the route to :basic by default" do
      SMSTradeRB.new.route.should == :basic
    end

    it "raises an exception on invalid routes" do
      expect {
        SMSTradeRB.new(:route => :eeek_invalid)
      }.to raise_error(SMSTradeRB::InvalidRoute)
    end
  end

  describe "additional options and attributes" do
    describe "debug" do
      it "it returns 1 if set to true" do
        SMSTradeRB.new(:debug => true).debug.should be(1)
      end

      it "it returns 0 if set to false" do
        SMSTradeRB.new(:debug => false).debug.should be(0)
      end
    end

    it "has a charset attribute" do
      SMSTradeRB.new.charset.should_not be_nil
    end
  end

  describe "#message" do
    it "returns an urlencoded string" do
      SMSTradeRB.new(:message => 'f$oo bar').message.should == 'f%24oo+bar'
    end
  end

  describe "#from" do
    it "returns an urlencoded string" do
      SMSTradeRB.new(:route => :gold, :from => 'f$oo bar').from.should == 'f%24oo+bar'
    end
  end

  describe "#send" do
    it "will encode and sanitize the phone number correctly" do
      app = SMSTradeRB::Server.new(:code => 999)
      Artifice.activate_with(app) do
        sms = SMSTradeRB.new(:route => :basic, :key => 'mykey', :debug => false)
        sms.send(:to => '+12 34-567', :message => 'my message')
      end

      app.params['to'].should == '+1234567'
    end

    it "will encode the message correctly" do
      app = SMSTradeRB::Server.new(:code => 999)
      Artifice.activate_with(app) do
        sms = SMSTradeRB.new(:route => :basic, :key => 'mykey', :debug => false)
        sms.send(:to => '+12 34-567', :message => 'my message $foo+bar')
      end

      app.params['to'].should == '+1234567'
    end

    context "success" do
      it "returns a response object" do
        Artifice.activate_with(SMSTradeRB::Server.new(:code => 999)) do
          sms = SMSTradeRB.new(:route => :basic, :key => 'mykey', :debug => false)
          sms.send(:to => '+1234', :message => 'my message').code.should be(999)
        end
      end
    end
  end
end


spec/smstraderb/server_spec.rb
require 'spec_helper'
require 'smstraderb/server'

describe SMSTradeRB::Server do
  def app
    SMSTradeRB::Server.new
  end

  before(:each) do
    @params = {
      :key => 'secret-key',
      :route => 'basic',
      :to => '123',
      :message => 'message text'
    }
  end

  describe "#new" do
    it "takes an options hash" do
      rack_app = SMSTradeRB::Server.new(:testop => 'foobar')
      rack_app[:testop].should == 'foobar'
    end
  end

  it "responds with the code in the :code option" do
    def app
      SMSTradeRB::Server.new(:code => 999)
    end
    get '/', @params do |response|
      response.body.should == '999'
    end
  end

  it "responds to the '/' url" do
    get '/' do |response|
      response.should be_ok
    end
  end

  it "responds with '10' if no recipient is set" do
    @params.delete(:to)
    get '/', @params do |response|
      response.body.should == '10'
    end
  end

  it "responds with '10' if the recipient format is wrong" do
    @params[:to] = 'abcde'
    get '/', @params do |response|
      response.body.should == '10'
    end
  end

  it "responds with '30' if no message text is set" do
    @params.delete(:message)
    get '/', @params do |response|
      response.body.should == '30'
    end
  end

  it "responds with '31' if the message type is invalid" do
    @params[:messagetype] = 'foobar'
    get '/', @params do |response|
      response.body.should == '31'
    end
  end

  it "responds with '40' if no route is set" do
    @params.delete(:route)
    get '/', @params do |response|
      response.body.should == '40'
    end
  end

  it "responds with '40' if an invalid route is set" do
    @params[:route] = 'foobar'
    get '/', @params do |response|
      response.body.should == '40'
    end
  end

  it "responds with '50' if no key is set" do
    @params.delete(:key)
    get '/', @params do |response|
      response.body.should == '50'
    end
  end

  describe "optional parameter" do
    describe "message_id" do
      it "responds with the message id in the second line" do
        @params[:message_id] = '1'
        get '/', @params do |response|
          response.body.should == "100\n123456789"
        end
      end
    end

    describe "cost" do
      it "responds with the costs in the third line" do
        @params[:cost] = '1'
        get '/', @params do |response|
          response.body.should == "100\n\n0.055"
        end
      end
    end

    describe "count" do
      it "responds with the sms count in the fourth line" do
        @params[:count] = '1'
        get '/', @params do |response|
          response.body.should == "100\n\n\n1"
        end
      end
    end
  end

  describe "call inspection" do
    it "provides access to the received params" do
      @rack_app = SMSTradeRB::Server.new
      def app
        @rack_app
      end
      get '/', @params

      @rack_app.params['message'].should == @params[:message]
    end
  end
end


lib/smstraderb.rb
require 'net/https'
require 'smstraderb/constants'
require 'smstraderb/response'

class SMSTradeRB
  class InvalidRoute < Exception; end
  class InvalidOption < Exception; end
  class InvalidFormat < Exception; end

  attr_reader :key, :message, :to, :from, :route, :debug, :charset

  def initialize(options = {})
    @charset = 'UTF-8'

    @key = options[:key]
    @message = options[:message]
    @to = check_phone_number(options[:to])
    @route = check_route(options[:route]) || SMSTradeRB::DEFAULT_ROUTE
    @from = options[:from]
    @debug = options[:debug] ? 1 : 0

    check_from_allowed(@from, @route)
    check_value_length(@from, 11) if @from

    @message = escape(@message)
    @from = escape(@from)
  end

  def send(options)
    @to = escape(check_phone_number(options[:to]))
    @message = escape(options[:message])

    uri = URI.parse(SMSTradeRB::HTTP_GATEWAY)
    http = Net::HTTP.new(uri.host, uri.port)
    #http.use_ssl = true if uri.scheme == "https"  # enable SSL/TLS

    response = http.start do
      http.request_get(build_request)
    end

    SMSTradeRB::Response.new(response.body)
  end

  private
    def build_request
      options = {
        :key => key,
        :message => message,
        :to => to,
        :route => route,
        :from => from,
        :debug => debug,
        :charset => charset
      }.map do |key, value|
        [key, value].join('=') if value
      end.compact

      "/?#{options.join('&')}"
    end

    def check_phone_number(number)
      if number and number.to_s !~ %r{^\+?[\s0-9\-\(\)]+$}
        raise SMSTradeRB::InvalidFormat, "Invalid phone number: #{number}"
      else
        sanitize_phone_number(number)
      end
    end

    def check_from_allowed(from, route)
      if from and ![:gold, :direct].include?(route)
        raise SMSTradeRB::InvalidOption, "Setting the from option is only allowed with a :gold or :direct route."
      end
    end

    def check_route(route)
      return nil unless route

      if SMSTradeRB::ROUTES.include?(route.to_sym)
        route
      else
        raise SMSTradeRB::InvalidRoute, "Route #{route} does not exist."
      end
    end

    def check_value_length(value, limit)
      if (bytesize(value) > 11)
        raise SMSTradeRB::InvalidFormat, "Value '#{value}' too long. (limit: #{limit})"
      end
    end

    def sanitize_phone_number(number)
      return nil unless number
      number.gsub(%r{[\s\-\(\)]+}, '')
    end

    # The following two methods came from rack-1.1.0.
    def escape(s)
      return nil unless s
      s.to_s.gsub(/([^ a-zA-Z0-9_.-]+)/n) {
        '%' + $1.unpack('H2'*bytesize($1)).join('%').upcase
      }.tr(' ', '+')
    end

    def bytesize(s)
      s.respond_to?(:bytesize) ? s.bytesize : s.size
    end
end


lib/smstraderb/response.rb
require 'smstraderb/constants'

class SMSTradeRB
  class InvalidResponse < Exception; end
  class UnknownResponse < Exception; end

  class Response
    def initialize(response)
      check_response_string(response)
      @code, @message_id, @cost, @count = response.split("\n")
      check_response_code
    end

    def code
      @code.to_i
    end

    def message_id
      blank?(@message_id) ? nil : @message_id
    end

    def cost
      blank?(@cost) ? nil : @cost.to_f
    end

    def count
      blank?(@count) ? nil : @count.to_i
    end

    def ok?
      code == 100
    end

    def error?
      !ok?
    end

    def response_message
      SMSTradeRB::RESPONSE_CODES[code]
    end

    private
      def check_response_string(value)
        if value.nil?
          raise SMSTradeRB::InvalidResponse, 'Response string is nil.'
        end
        if value.empty?
          raise SMSTradeRB::InvalidResponse, 'Response string is empty.'
        end
        if value !~ %r{^\d+}
          raise SMSTradeRB::InvalidResponse, "Invalid response string: #{value}."
        end
      end

      def check_response_code
        unless SMSTradeRB::RESPONSE_CODES.key?(code)
          raise SMSTradeRB::UnknownResponse
        end
      end

      def blank?(value)
        value.nil? || value.empty? ? true : false
      end
  end
end


lib/smstraderb/server.rb
require 'smstraderb/constants'

# This class tries to emulate the smstrade.de HTTP API v2.2.
# It provides a valid Rack endpoint.

class SMSTradeRB
  class Server
    attr_reader :params

    def initialize(options = {})
      @options = options
    end

    def [](key)
      @options[key]
    end

    def call(env)
      @params = Rack::Request.new(env).params

      # The route param is mandatory.
      unless @params['route']
        return respond_with('40')
      end

      # There are only a few valid route values.
      unless SMSTradeRB::ROUTES.include?(@params['route'].to_sym)
        return respond_with('40')
      end

      # The to params is mandatory.
      unless @params['to']
        return respond_with('10')
      end

      # The to params needs to be in a valid format.
      unless @params['to'] =~ /^\+?\d+/
        return respond_with('10')
      end

      # The key param is mandatory.
      unless @params['key']
        return respond_with('50')
      end

      # The message params is mandatory.
      unless @params['message']
        return respond_with('30')
      end

      # Check if the messagetype is valid.
      if @params['messagetype'] and !SMSTradeRB::MESSAGE_TYPES.include?(@params['messagetype'].to_sym)
        return respond_with('31')
      end

      # Everything is fine so far. We return 100.
      ret = [@options[:code] || 100]

      # Some optional parameters which modify the return value.
      if @params['message_id'] == '1'
        ret[1] = '123456789'
      end

      if @params['cost'] == '1'
        ret[2] = '0.055'
      end

      if @params['count'] == '1'
        ret[3] = '1'
      end

      respond_with(*ret)
    end

    private
      def respond_with(*values)
        [200, {'Content-Type' => 'text/plain'}, [values.join("\n")]]
      end
  end
end
